crates/dbsp/src/circuit/circuit_builder.rs

```diff
@@ -564,327 +574,978 @@ impl Edge {
+/// node1 to node2 indicates that the output stream of node1
+/// is connected to an input of node2.
+struct CircuitInner<P>
+where
+    P: WithClock,
+{
+    parent: P,
+    root_scope: Scope,
+    // Circuit's node id within the parent circuit.
+    node_id: NodeId,
+    global_node_id: GlobalNodeId,
+    nodes: Vec<Box<dyn Node>>,
+    edges: Vec<Edge>,
+    circuit_event_handlers: CircuitEventHandlers,
+    scheduler_event_handlers: SchedulerEventHandlers,
+    store: CircuitCache,
+}
+
+impl<P> CircuitInner<P>
+where
+    P: WithClock,
+{
+    fn new(
+        parent: P,
+        root_scope: Scope,
+        node_id: NodeId,
+        global_node_id: GlobalNodeId,
+        circuit_event_handlers: CircuitEventHandlers,
+        scheduler_event_handlers: SchedulerEventHandlers,
+    ) -> Self {
+        Self {
+            parent,
+            root_scope,
+            node_id,
+            global_node_id,
+            nodes: Vec::new(),
+            edges: Vec::new(),
+            circuit_event_handlers,
+            scheduler_event_handlers,
+            store: TypedMap::new(),
+        }
+    }
+
+    fn add_edge(&mut self, edge: Edge) {
+        self.edges.push(edge);
+    }
+
+    fn add_node<N>(&mut self, node: N)
+    where
+        N: Node + 'static,
+    {
+        self.nodes.push(Box::new(node) as Box<dyn Node>);
+    }
+
+    fn clear(&mut self) {
+        self.nodes.clear();
+        self.edges.clear();
+        self.store.clear();
+    }
+
+    fn register_circuit_event_handler<F>(&mut self, name: &str, handler: F)
+    where
+        F: Fn(&CircuitEvent) + 'static,
+    {
+        self.circuit_event_handlers.borrow_mut().insert(
+            name.to_string(),
+            Box::new(handler) as Box<dyn Fn(&CircuitEvent)>,
+        );
+    }
+
+    fn unregister_circuit_event_handler(&mut self, name: &str) -> bool {
+        self.circuit_event_handlers
+            .borrow_mut()
+            .remove(name)
+            .is_some()
+    }
+
+    fn register_scheduler_event_handler<F>(&mut self, name: &str, handler: F)
+    where
+        F: FnMut(&SchedulerEvent<'_>) + 'static,
+    {
+        self.scheduler_event_handlers.borrow_mut().insert(
+            name.to_string(),
+            Box::new(handler) as Box<dyn FnMut(&SchedulerEvent<'_>)>,
+        );
+    }
+
+    fn unregister_scheduler_event_handler(&mut self, name: &str) -> bool {
+        self.scheduler_event_handlers
+            .borrow_mut()
+            .remove(name)
+            .is_some()
+    }
+
+    fn log_circuit_event(&self, event: &CircuitEvent) {
+        for (_, handler) in self.circuit_event_handlers.borrow().iter() {
+            handler(event)
+        }
+    }
+
+    fn log_scheduler_event(&self, event: &SchedulerEvent<'_>) {
+        for (_, handler) in self.scheduler_event_handlers.borrow_mut().iter_mut() {
+            handler(event)
+        }
+    }
+
+    fn fixedpoint(&self, scope: Scope) -> bool {
+        self.nodes.iter().all(|node| {
```

Comment:
could this be done incrementally too? maybe it's not worth it

Reply:
You mean, have nodes notify the circuit when their fixedpoint status changes?  Yeah, probably not worth it at the moment.